client/cgo/coniks.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import "C"
+
+import (
+	"bytes"
+	"unsafe"
+
+	"github.com/coniks-sys/coniks-go/crypto"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+// main is required to build a shared library, but does nothing
+func main() {}
+
+//export cgoEd25519SignatureVerify
+func cgoEd25519SignatureVerify(pk unsafe.Pointer, pkSize C.int,
+	message unsafe.Pointer, size C.int,
+	sig unsafe.Pointer, sigSize C.int) C.int {
+	if int(pkSize) < crypto.PublicKeySize {
+		return 0
+	}
+	pkBytes := C.GoBytes(pk, pkSize)
+	messageBytes := C.GoBytes(message, size)
+	sigBytes := C.GoBytes(sig, sigSize)
+	key := crypto.VerifKey(pkBytes)
+	if key.Verify(messageBytes, sigBytes) {
+		return 1
+	}
+	return 0
+}
+
+//export cgoVrfVerify
+func cgoVrfVerify(pk unsafe.Pointer, pkSize C.int,
+	m unsafe.Pointer, size C.int,
+	index unsafe.Pointer, indexSize C.int,
+	proof unsafe.Pointer, proofSize C.int) C.int {
+	if int(pkSize) < vrf.PublicKeySize ||
+		int(indexSize) < vrf.Size ||
+		int(proofSize) < vrf.ProofSize {
+		return 0
+	}
+	pkBytes := C.GoBytes(pk, pkSize)
+	mBytes := C.GoBytes(m, size)
+	vrfBytes := C.GoBytes(index, indexSize)
+	proofBytes := C.GoBytes(proof, proofSize)
+	if vrf.Verify(pkBytes, mBytes, vrfBytes, proofBytes) {
```

Comment:
> pad.key is sometimes used for signing data, sometimes used for VRF functions.

`pad.key` is just used for signing data.  `policies.vrfPrivKey` is used for VRF functions. So I think it's not really a problem here. I renamed `pad.key` to `pad.signKey` for clarifying.

> What I think we should do is not expose vrf at all to pad or anywhere else. We should have one API in crypto that abstracts this all away, and crypto will call vrf where necessary.
So, getting to L48 here, pkBytes should be cast to a VerifKey (or whatever public key type we end up on), and then you'd call pk.VerifyVRF or something like that.

Agree! But what if we use 2 different types of verification key for signature and VRF in the future? Then we probably cannot have only one VerifKey type. Should we have some APIs named: `crypto.SignatureGenerateKey`, `crypto.VrfGenerateKey` and key types: `SignatureVerifKey`, `VrfVerifKey`, then I'd call `pk.Verify`. What do you think?